File: Python/Decorator/Decorator.py
# -*- coding:utf-8 -*-

from MobilePhone import *

class Decorator(MobilePhone):

    def __init__(self, phone):
        self.__phone = phone

    def ShowDecorate(self):
        self.__phone.ShowDecorate()

class Decorator1(Decorator):

    def __init__(self, phone):
        Decorator.__init__(self, phone)

    def ShowDecorate(self):
        Decorator.ShowDecorate(self)
        self.__AddDecorate()

    def __AddDecorate(self):
        print "增加Decorator1"

class Decorator2(Decorator):

    def __init__(self, phone):
        Decorator.__init__(self, phone)


    def ShowDecorate(self):
        Decorator.ShowDecorate(self)
        self.__AddDecorate()


    def __AddDecorate(self):
        print "增加Decorator2"

File: CPlusPlus/Decorator/src/Decorator.cpp
/*************************************************************************
    > File Name: Decorator.cpp
    > Author: ma6174
    > Mail: ma6174@163.com 
    > Created Time: 2017年03月02日 星期四 14时28分15秒
 ************************************************************************/

#include"Decorator.h"

Decorator::Decorator(MobilePhone* tempPhone):MobilePhone(tempPhone->GetPhoneName()),phone(tempPhone)
{
	;
}

Decorator::~Decorator()
{
	;
}

void Decorator::ShowDecorate()
{
	phone->ShowDecorate();
}
